src/vtarget-out/inv-syn/inv_abc_parse.cc

```diff
@@ -0,0 +1,639 @@
+              } else 
+                remove_this_cube = true;
+            }else {
+              literal = (neg ? "~" : "") +  ref_val_name;
+              cl.push_back(InvariantInCnf::literal(bool(neg),lit.first,lit.second));
+            }
+          }
+        }
+
+ // refering to outside var
+        if(cube.empty())
+          cube = literal;
+        else
+          cube = cube + " & "  + literal;
+      } // end of a cube
+      // cube = "( " +  cube + ") "
+
+      // void the cube if abnormal
+      if( remove_this_cube) // cube_has_abnormal_var ||
+        cube = "1'b0" ; //cube = "1'b0"; // under-approximate it : total under-approximate
+      else {// if at least the cube has a normal var, it is a normal cube
+        has_a_normal_cube = true;
+        // only if this cube is not removed, we will add it 
+        if (parse_result.empty())
+          parse_result =  "(" + cube + ")";
+        else
+          parse_result = parse_result + " | (" + cube + ")";
+        inv_cnf.InsertClauseNewerFromReference(cl, ref_cnf);
+      }
+    } // deal with a line ( a clause )
+    ILA_ERROR_IF(has_a_normal_cube == false) << "No normal is deduced.";
+    ILA_ERROR_IF(parse_result.empty()) << "No invariants are extracted!";
+    if (!has_a_normal_cube || parse_result.empty() ) {
+      parse_result = "";
+      parse_succeed = false;
+    }
+    
+    parse_result = "~("  + parse_result + ")";
+    parse_result = ReplaceAll(parse_result, "__all_assumed_reg__[0:0]", "1'b0");
+  } // parse abc file
+
+} // parseAigerResultWoGLA
+
+
+void AbcInvariantParser::parseAigerResultWithGLA(
+    const std::string & aig_map_fn,
+    const std::string & abc_result_fn,
+    const std::string & gla_map_fn,
+    InvariantInCnf & inv_cnf,
+    const InvariantInCnf & ref_cnf,
+    const std::string & blif_fn_name) {
+    
+
+  std::vector<std::string> aig_state_order;
+  std::vector<std::pair<std::string, unsigned>> aig_literal;
+  std::set<std::string> blif_valid_state_names;
+  std::set<int> invlatches;
+
+  { // read blif for valid state names
+    std::ifstream fin(blif_fn_name);
+    if (! fin.is_open()) {
+      ILA_ERROR << "Cannot read from: " <<blif_fn_name;
+      return;
+    }
+    std::string line;
+    while(std::getline(fin,line)) {
+      if (line.find(".latch ") == 0) {
+        auto vec = SplitSpaceTabEnter(line);
+        blif_valid_state_names.insert(vec.at(2));
+      }
+    }
+  } // read blif for valid state names
+
+  { // read aiger map
+    std::ifstream fin(aig_map_fn);
+    if (! fin.is_open()) {
+      ILA_ERROR << "Cannot read from: " <<aig_map_fn;
+      return;
+    }
+    std::string line;
+    while(std::getline(fin,line)) {
+      auto vec = SplitSpaceTabEnter(line);
+      // only focus on the registers
+      if (line.find("latch ") == 0 || line.find("invlatch ") == 0) {
+        auto vec = SplitSpaceTabEnter(line);
+        unsigned long long latch_no = StrToULongLong(vec.at(1));
+        if (line.find("invlatch ") == 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Inefficient usage of string::find() in condition; string::starts_with() could be faster.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)